Stack picker and preview vertically if too narrow

zqianem opened the pull request:
I couldn't get #7783 to work well for me so I rewrote it.

@pascalkuthe if you're still interested now that #9647 has landed.

I tried @archseer's [suggestion](https://github.com/helix-editor/helix/pull/7783#issuecomment-2273713481) but in practice, I didn't like it for shorter terminals that you would still want a vertical picker for because they were also narrow.

Looks like this:
<img width="1007" height="1390" alt="file picker open with preview shown stacked below it" src="https://github.com/user-attachments/assets/6c818024-e745-43c6-b728-d25da12543df" />


At `MIN_AREA_HEIGHT_FOR_PREVIEW`, eight entries are shown:
<img width="1007" height="651" alt="file picker open with eight search results before a preview with height 24" src="https://github.com/user-attachments/assets/65040505-9b93-4367-8422-07b985f9f22a" />

zqianem committed:
Stack picker and preview vertically if too narrow

Co-authored-by: Jesse Luehrs <doy@tozt.net>

zqianem committed:
Refactor common logic into helper function

zqianem committed:
Fix `cargo clippy` error

zqianem committed:
Use existing `impl` for helper function

alexpyzhianov commented on helix-term/src/ui/picker.rs:
```diff
@@ -984,31 +985,58 @@ impl<T: 'static + Send + Sync, D: 'static + Send + Sync> Picker<T, D> {
+        let render_preview = self.show_preview
+            && self.file_fn.is_some()
+            && area.width >= MIN_AREA_WIDTH_FOR_PREVIEW
+            && area.height >= MIN_AREA_HEIGHT_FOR_PREVIEW;
```
Found a small bug, playing with your patch locally (great work btw!)

I think this should be an OR condition rather than AND, so 
```
(area.width >= MIN_AREA_WIDTH_FOR_PREVIEW || area.height >= MIN_AREA_HEIGHT_FOR_PREVIEW)
```

zqianem replied:
Fixed by 4ff2160

alexpyzhianov commented:
Great work @zqianem ! Much appreciated! Looking forward to seeing this in production!

zqianem committed:
Keep the preview visible at very narrow widths

Now, the `MIN_AREA_WIDTH_FOR_PREVIEW` just determines when to switch to
stack layout — there's no reason to hide the preview once we are in a
stack layout just because the editor is very narrow.

zqianem commented:
@the-mikedavis whoops, I earlier merged an outdated master branch which was what was causing the build failures; fixed now.

zqianem commented:
Rebased and updated the screenshots to reflect what this looks like after #12311
